include/fiction/algorithms/physical_design/graph_oriented_layout_design.hpp

```diff
@@ -0,0 +1,1574 @@
+    [[nodiscard]] node index_to_node(const uint32_t index) const
+    {
+        return topo_order.at(index);
+    }
+    /*! \brief Reimplementation of `foreach_node`. */
+    template <typename Fn>
+    void foreach_node(Fn&& fn) const
+    {
+        mockturtle::detail::foreach_element(topo_order.cbegin(), topo_order.cend(), std::forward<Fn>(fn));
+    }
+    /*! \brief Reimplementation of `foreach_gate`. */
+    template <typename Fn>
+    void foreach_gate(Fn&& fn) const
+    {
+        const uint32_t offset = 1u + this->num_pis() +
+                                (this->get_node(this->get_constant(true)) != this->get_node(this->get_constant(false)));
+        mockturtle::detail::foreach_element(topo_order.cbegin() + offset, topo_order.cend(), std::forward<Fn>(fn));
+    }
+    /*! \brief Implementation of `foreach_gate` in reverse topological order. */
+    template <typename Fn>
+    void foreach_gate_reverse(Fn&& fn) const
+    {
+        const uint32_t offset = 1u + this->num_pis() +
+                                (this->get_node(this->get_constant(true)) != this->get_node(this->get_constant(false)));
+        mockturtle::detail::foreach_element(topo_order.crbegin(), topo_order.crend() - offset, std::forward<Fn>(fn));
+    }
+
+    void update_topo()
+    {
+        this->incr_trav_id();
+        this->incr_trav_id();
+        topo_order.reserve(this->size());
+
+        /* constants and PIs */
+        const auto c0 = this->get_node(this->get_constant(false));
+        topo_order.push_back(c0);
+        this->set_visited(c0, this->trav_id());
+
+        if (const auto c1 = this->get_node(this->get_constant(true)); this->visited(c1) != this->trav_id())
+        {
+            topo_order.push_back(c1);
+            this->set_visited(c1, this->trav_id());
+        }
+
+        Ntk::foreach_ci([this](const auto& n) { create_topo_rec(n); });
+    }
+
+  private:
+    void create_topo_rec(const node& n)
+    {
+        /* is permanently marked? */
+        if (this->visited(n) == this->trav_id())
+        {
+            return;
+        }
+
+        bool not_visited = false;
+
+        this->foreach_fanin(n,
+                            [this, &not_visited](const signal& f)
+                            {
+                                if (this->visited(this->get_node(f)) != this->trav_id())
+                                {
+                                    not_visited = true;
+                                };
+                            });
+
+        if (not_visited)
+        {
+            return;
+        }
+
+        /* mark node n permanently */
+        this->set_visited(n, this->trav_id());
+
+        /* visit node */
+        topo_order.push_back(n);
+
+        /* mark children */
+        this->foreach_fanout(n, [this](const auto& fo) { create_topo_rec(fo); });
+    }
+
+    std::vector<node> topo_order;
+};
+/**
+ * Enum to specify which tiles should be checked if they are empty.
+ */
+enum class check_if_empty
```

Comment:
I'm not super happy with the name and docstring of this `enum class`. I can infer from neither what this is used for. Can we find a more descriptive naming and documentation here?